### DISCUSSION TITLE
Can't find devices.

### DISCUSSION THREAD

**Discussion by mthnzbk** [dotnet-bluetooth-le/dotnet-bluetooth-le#766] [Q&A] [answered]
I don't get any error in the code below, but the StartScanningForDevicesAsync() method does not find open bluetooth devices. So Adapter.DeviceDiscovered += async (s, a) => {} I don't get any output in this part. Nor error.

## Configuration

**Version of the Plugin:** 3.0.0 

**Platform:**Android 10-11 

**Device:** Xiaomi Note 7
`using Plugin.BLE;
using Plugin.BLE.Abstractions.Contracts;
using System.Collections.ObjectModel;
using System.Diagnostics;

namespace Kumanda
{
    public partial class MainPage : ContentPage
    {
        readonly IBluetoothLE Ble = CrossBluetoothLE.Current;
        readonly IAdapter Adapter = CrossBluetoothLE.Current.Adapter;

        public ObservableCollection<String> Devices;
        public ObservableCollection<IDevice> Device;

        public MainPage()
        {
            InitializeComponent();

            Ble.StateChanged += (s, e) =>
            {
                Debug.WriteLine("State: " + e.NewState);
            };

            Adapter.DeviceDiscovered += async (s, a) =>
            {
                Debug.WriteLine("Device: " + a.Device.ToString());
                try
                {
                    Devices.Add(a.Device.ToString());
                    Device.Add(a.Device);
                    await Adapter.ConnectToDeviceAsync(a.Device);
                    var service = await a.Device.GetServicesAsync();
                    Debug.WriteLine(service.ToString());
                    Debug.WriteLine(Devices + " " + Device);
                }
                catch (Exception ex)
                {
                    Debug.WriteLine(ex);
                }
                
            };

            Adapter.DeviceBondStateChanged += (s, e) =>
            {
                Debug.WriteLine(e.State + " " + e.Address);
            };

            Adapter.DeviceConnected += (s, e) =>
            {
                Debug.WriteLine(e.Device.ToString());
            };
        }

        protected override void OnAppearing()
        {
            base.OnAppearing();
            
        }

        private async void OnCounterClicked(object sender, EventArgs e)
        {
            Debug.WriteLine(Ble.State);
            try
            {
                await Adapter.StartScanningForDevicesAsync();
            }
            catch (Exception ex)
            {
                Debug.WriteLine(ex);
            }
        }
    }
}`

**Comment by smsissuechecker** [comment]
Hi @mthnzbk,

I'm the friendly issue checker.
It seems like (50.00 %) you haven't used our issue template :cry: I think it is very frustrating for the repository owners, if you ignore them.

If you think it's fine to make an exception, just ignore this message.
**But if you think it was a mistake to delete the template, please close the issue and create a new one.**

Thanks!

**Comment by janusw** [comment]
> I don't get any error in the code below, but the StartScanningForDevicesAsync() method does not find open bluetooth devices. So Adapter.DeviceDiscovered += async (s, a) => {} I don't get any output in this part. Nor error.

Possibly you're failing to request permissions properly? See https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le#permissions

You need to add the relevant permissions to your manifest and also request them at runtime.

Which SDK level are you targeting? Do you use Xamarin or MAUI?

**Comment by mthnzbk** [comment]
`<uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
<uses-permission android:name="android.permission.INTERNET" />
<uses-permission android:name="android.permission.BLUETOOTH" />
<uses-permission android:name="android.permission.BLUETOOTH_ADMIN" />
<uses-permission android:name="android.permission.BLUETOOTH_CONNECT" />
<uses-permission android:name="android.permission.BLUETOOTH_ADVERTISE" />
<uses-permission android:name="android.permission.BLUETOOTH_SCAN" />
<uses-permission android:name="android.permission.READ_USER_DICTIONARY" />
<uses-permission android:name="android.permission.VIBRATE" />
<uses-permission android:name="android.permission.WRITE_USER_DICTIONARY" />
<uses-sdk android:minSdkVersion="26" android:targetSdkVersion="33" />`

`protected override void OnCreate(Bundle savedInstanceState)
{
    base.OnCreate(savedInstanceState);
    if (ContextCompat.CheckSelfPermission(this, Manifest.Permission.BluetoothScan) != Permission.Granted)
    {
        ActivityCompat.RequestPermissions(this, new[] {Manifest.Permission.BluetoothScan }, 0);
    }

    if (ContextCompat.CheckSelfPermission(this, Manifest.Permission.BluetoothAdvertise) != Permission.Granted)
    {
        ActivityCompat.RequestPermissions(this, new[] { Manifest.Permission.BluetoothAdvertise }, 0);
    }
    if (ContextCompat.CheckSelfPermission(this, Manifest.Permission.BluetoothConnect) != Permission.Granted)
    {
        ActivityCompat.RequestPermissions(this, new[] { Manifest.Permission.BluetoothConnect }, 0);
    }
    if (ContextCompat.CheckSelfPermission(this, Manifest.Permission.BluetoothAdmin) != Permission.Granted)
    {
        ActivityCompat.RequestPermissions(this, new[] { Manifest.Permission.BluetoothAdmin }, 0);
    }
}`

.net 7 Maui

**Reply by Moes81** [reply]
Hm, same problem on Android 12 (Pixel 4).

these are my Permissions:
```
public class Android11BluetoothPermission : Microsoft.Maui.ApplicationModel.Permissions.BasePlatformPermission
{
#if ANDROID
    public override (string androidPermission, bool isRuntime)[] RequiredPermissions =>
        new List<(string permission, bool isRuntime)>
        {
            ("android.permission.ACCESS_COARSE_LOCATION", true),
            ("android.permission.ACCESS_FINE_LOCATION", true),
            ("android.permission.BLUETOOTH", true),
            ("android.permission.BLUETOOTH_ADMIN", true)
        }.ToArray();
#endif
}

public class Android12BluetoothPermission : Microsoft.Maui.ApplicationModel.Permissions.BasePlatformPermission
{
#if ANDROID
    public override (string androidPermission, bool isRuntime)[] RequiredPermissions =>
        new List<(string permission, bool isRuntime)>
        {
            ("android.permission.BLUETOOTH_SCAN", true),
            ("android.permission.BLUETOOTH_CONNECT", true)
        }.ToArray();
#endif
}

```

Permissions are requested properly, are granted, but still no scan results. :-( Not sure what to do or how to debug. 
This is the start scan method, using Rx.Net to publish the list of found devices. However the `OnAdapterOnDeviceDiscovered` is never called.

```
public IObservable<IEnumerable<DeviceCandidate>> StartBluetoothScan()
    {
        if (_scanObservable != null)
        {
            return _scanObservable;
        }

        var observable = Observable.Create<IEnumerable<DeviceCandidate>>(observer =>
        {
            _scanObserver = observer;
            var deviceCandidates = new List<DeviceCandidate>();

            _adapter.ScanMode = ScanMode.LowLatency;
            _adapter.ScanMatchMode = ScanMatchMode.AGRESSIVE;
            _adapter.ScanTimeout = 30000;
            _adapter.DeviceAdvertised += OnAdapterOnDeviceDiscovered;
            _adapter.DeviceDiscovered += OnAdapterOnDeviceDiscovered;

            return Disposable.Create(() =>
            {
                _adapter.DeviceAdvertised -= OnAdapterOnDeviceDiscovered;
                _adapter.DeviceDiscovered -= OnAdapterOnDeviceDiscovered;
                Task.Run(async () => await StopBluetoothScanAsync());
            });

            void OnAdapterOnDeviceDiscovered(object? _, DeviceEventArgs args)
            {
                deviceCandidates.Add(new DeviceCandidate(args.Device.Id, args.Device.Name));
                observer.OnNext(deviceCandidates.DistinctBy(candidate => candidate.Id));
            }
        });

        _scanObservable = observable;
        Task.Run(async () => await _adapter.StartScanningForDevicesAsync(allowDuplicatesKey: true)
        );
        return observable;
    }
```

Btw. How to I remove the time out so I can scan forever?

**Reply by janusw** [reply]
See my comment below: https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/discussions/766#discussioncomment-7407132

**Reply by janusw** [reply]
> Btw. How to I remove the time out so I can scan forever?

simply set `CrossBluetoothLE.Current.Adapter.ScanTimeout` to a very large value

**Reply by Moes81** [reply]
I'm now able to get scan results with the following configuration:
```
//AndroidManifest.xml
  <uses-feature android:name="android.hardware.bluetooth_le" android:required="true"/>

  <!-- Request legacy Bluetooth + location permissions on older devices. -->
  <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION" android:maxSdkVersion="30"/>
  <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION" android:maxSdkVersion="30"/>
  <uses-permission android:name="android.permission.BLUETOOTH"/>
  <uses-permission android:name="android.permission.BLUETOOTH_ADMIN"/>

  <uses-permission android:name="android.permission.BLUETOOTH_SCAN" android:usesPermissionFlags="neverForLocation"/>
  <uses-permission android:name="android.permission.BLUETOOTH_CONNECT"/>

public class Android12BluetoothPermission : Microsoft.Maui.ApplicationModel.Permissions.BasePlatformPermission
{
#if ANDROID
    public override (string androidPermission, bool isRuntime)[] RequiredPermissions =>
        new List<(string permission, bool isRuntime)>
        {
            ("android.permission.BLUETOOTH_SCAN", true),
            ("android.permission.BLUETOOTH_CONNECT", true)
        }.ToArray();
#endif
}

public class Android11BluetoothPermission : Microsoft.Maui.ApplicationModel.Permissions.BasePlatformPermission
{
#if ANDROID
    public override (string androidPermission, bool isRuntime)[] RequiredPermissions =>
        new List<(string permission, bool isRuntime)>
        {
            ("android.permission.ACCESS_COARSE_LOCATION", true),
            ("android.permission.ACCESS_FINE_LOCATION", true)
        }.ToArray();
#endif
}
```

**Reply by janusw** [reply]
Great! I assume you fixed the problem by adding the `neverForLocation` flag in the manifest.

**Comment by janusw** [comment, marked as answer]
From https://developer.android.com/develop/connectivity/bluetooth/bt-permissions:

> If your app uses Bluetooth scan results to derive physical location, declare the [ACCESS_FINE_LOCATION](https://developer.android.com/reference/android/Manifest.permission#ACCESS_FINE_LOCATION) permission. Otherwise, you can [strongly assert that your app doesn't derive physical location](https://developer.android.com/develop/connectivity/bluetooth/bt-permissions#assert-never-for-location).

**Comment by mthnzbk** [comment]
I also added the missing permission, but again when the device search method works, it cannot find the pc and the other phone.

**Reply by janusw** [reply]
Well, hard to say what exactly is wrong, without a proper reproducer. I'd still bet on some kind of permission problem. We have a MAUI sample client that can act as a blueprint for your own client.

**Reply by janusw** [reply]
Or are you saying the scanning as such works now, and you only miss specific devices?
